<?xml version="1.0"?>
<robot name="robot_arm">

  <link name="base_link">
    <visual>
      <origin xyz="0 0 0.050" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.027" length="0.100" />
      </geometry>
      <material name="red">
        <color rgba=".8 0 0 1" />
      </material>
    </visual>
  </link>

  <joint name="joint0" type="revolute">
    <parent link="base_link"/>
    <child link="link1"/>
    <origin xyz="0 0 0.100" rpy="0 0 0" />
    <axis xyz="0 0 1" />
    <limit lower="-1.57" upper="1.57" effort="0.5293" velocity="6.159937255" />
  </joint>

  <link name="link1">
    <visual>
      <origin xyz="0 0 0.006" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.027" length="0.012" />
      </geometry>
      <material name="green">
        <color rgba="0 .8 0 1" />
      </material>
    </visual>
  </link>

  <joint name="joint1" type="revolute">
    <parent link="link1"/>
    <child link="link2"/>
    <origin xyz="0 0 0.012" rpy="0 0 0" />
    <axis xyz="0 -1 0" />
    <limit lower="-1.57" upper="1.57" effort="0.5293" velocity="6.159937255" />
  </joint>

  <link name="link2">
    <visual>
      <origin xyz="0 0 0.052" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.027" length="0.104" />
      </geometry>
      <material name="blue">
        <color rgba="0 0 .8 1" />
      </material>
    </visual>
  </link>

  <joint name="joint2" type="revolute">
    <parent link="link2"/>
    <child link="link3"/>
    <origin xyz="0 0 0.104" rpy="0 0 0" />
    <axis xyz="0 1 0" />
    <limit lower="-1.57" upper="1.57" effort="0.5293" velocity="6.159937255" />
  </joint>

  <link name="link3">
    <visual>
      <origin xyz="0 0 0.049" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.027" length="0.098" />
      </geometry>
      <material name="yellow">
        <color rgba=".8 .8 0 1" />
      </material>
    </visual>
  </link>

  <joint name="joint3" type="revolute">
    <parent link="link3"/>
    <child link="link4"/>
    <origin xyz="0 0 0.098" rpy="0 0 0" />
    <axis xyz="0 1 0" />
    <limit lower="-1.57" upper="1.57" effort="0.5293" velocity="6.159937255" />
  </joint>

  <link name="link4">
    <visual>
      <origin xyz="-0.014 0 0.028" rpy="0 0 0" />
      <geometry>
        <box size="0.028 0.054 0.056" />
      </geometry>
      <material name="cyan">
        <color rgba="0 .8 .8 1" />
      </material>
    </visual>
  </link>

  <joint name="joint4" type="revolute">
    <parent link="link4"/>
    <child link="gripper_base"/>
    <origin xyz="-0.028 0.012 0.056" rpy="0 0 0" />
    <axis xyz="0 0 1" />
    <limit lower="-1.57" upper="1.57" effort="0.5293" velocity="6.159937255" />
  </joint>

  <link name="gripper_base">
    <visual>
      <origin xyz="0 0 0.014" rpy="0 0 0" />
      <geometry>
        <cylinder radius="0.010" length="0.028" />
      </geometry>
      <material name="purple">
        <color rgba=".8 0 .8 1" />
      </material>
    </visual>
  </link>

  <joint name="left_gripper_joint" type="revolute">
    <parent link="gripper_base"/>
    <child link="left_gripper"/>
    <origin xyz="0 0.014 0.028" rpy="-1.0 0 0" />
    <axis xyz="1 0 0" />
    <limit lower="0" upper="1.0" effort="1000.0" velocity="0.5"/>
  </joint>

  <link name="left_gripper">
    <visual>
      <origin rpy="0 -1.57 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://urdf_tutorial/meshes/l_finger.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="left_tip_joint" type="fixed">
    <parent link="left_gripper"/>
    <child link="left_tip"/>
  </joint>

  <link name="left_tip">
    <visual>
      <origin rpy="0 -1.57 0" xyz="0.00495 0 0.09137"/>
      <geometry>
        <mesh filename="package://urdf_tutorial/meshes/l_finger_tip.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="right_gripper_joint" type="revolute">
    <parent link="gripper_base"/>
    <child link="right_gripper"/>
    <origin xyz="0 -0.014 0.028" rpy="1.0 0 0" />
    <limit lower="0" upper="1.0" effort="1000.0" velocity="0.5"/>
    <mimic joint="left_gripper_joint" multiplier="-1" offset="0" />
  </joint>

  <link name="right_gripper">
    <visual>
      <origin rpy="-3.1415 -1.57 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://urdf_tutorial/meshes/l_finger.dae"/>
      </geometry>
    </visual>
  </link>

  <joint name="right_tip_joint" type="fixed">
    <parent link="right_gripper"/>
    <child link="right_tip"/>
  </joint>

  <link name="right_tip">
    <visual>
      <origin rpy="-3.1415 -1.57 0" xyz="0.00495 0 0.09137"/>
      <geometry>
        <mesh filename="package://urdf_tutorial/meshes/l_finger_tip.dae"/>
      </geometry>
    </visual>
  </link>

</robot>
